gflownet/envs/cube.py

```diff
@@ -0,0 +1,1317 @@
+        - 2 : whether EOS action is invalid. EOS is valid from any state, except the
+          source state or if done is True.
+        """
+        state = self._get_state(state)
+        done = self._get_done(done)
+        mask_dim = 3
+        # If done, the entire mask is True (all actions are "invalid" and no special
+        # cases)
+        if done:
+            return [True] * mask_dim
+        mask = [False] * mask_dim
+        # If the state is not the source state, EOS is invalid
+        if state == self.source:
+            mask[2] = True
+        # If the state is not the source, indicate not special case (True)
+        else:
+            mask[1] = True
+        # If the value of any dimension is greater than 1 - min_incr, then continuous
+        # actions are invalid (True).
+        if any([s > 1 - self.min_incr for s in state]):
+            mask[0] = True
+        return mask
+
+    def get_mask_invalid_actions_backward(self, state=None, done=None, parents_a=None):
+        """
+        The action space is continuous, thus the mask is not only of invalid actions as
+        in discrete environments, but also an indicator of "special cases", for example
+        states from which only certain actions are possible.
+
+        In order to approximately stick to the semantics in discrete environments,
+        where the mask is of "invalid" actions, that is the value is True if an action
+        is invalid, the mask values of special cases are True if the special cases they
+        refer to are "invalid". In other words, the values are False if the state has
+        the special case.
+
+        The backward mask has the following structure:
+
+        - 0 : whether a continuous action is invalid. True if the value at any
+          dimension is smaller than min_incr, or if done is True. False otherwise.
+        - 1 : special case when back-to-source action is the only possible action.
+          False if any dimension is smaller than min_incr, True otherwise.
+        - 2 : whether EOS action is invalid. False only if done is True, True
+          (invalid) otherwise.
+        """
+        state = self._get_state(state)
+        done = self._get_done(done)
+        mask_dim = 3
+        mask = [True] * mask_dim
+        # If done, only valid action is EOS.
+        if done:
+            mask[2] = False
+            return mask
+        # If any dimension is smaller than m, then back-to-source action is the only
+        # possible actiona.
+        if any([s < self.min_incr for s in state]):
+            mask[1] = False
+            return mask
+        # Otherwise, continuous actions are valid
+        mask[0] = False
+        return mask
+
+    def get_parents(
+        self, state: List = None, done: bool = None, action: Tuple[int, float] = None
+    ) -> Tuple[List[List], List[Tuple[int, float]]]:
+        """
+        Defined only because it is required. A ContinuousEnv should be created to avoid
```

Comment:
I agree, but is it a priority right now?

Reply:
Hell no, it's just a way for me to acknowledge that I also think it would be a good idea to have one. One day. When we retire, start living on a farm, and won't have any deadlines.